[JENKINS-55813] Check user validity in impersonate

jvz opened the pull request:
This ensures that user accounts used for performing actions are valid in all call sites.

See [JENKINS-55813](https://issues.jenkins-ci.org/browse/JENKINS-55813).

I'm opening this initially to test out some security realm changes elsewhere to validate this as including tests here would require extensive mocking.

### Proposed changelog entries

* JENKINS-55813, Security realms that provide account status checks are now properly enforced in unattended code that executes on behalf of a user.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

jvz committed:
[JENKINS-55813] Check user validity in impersonate

This ensures that user accounts used for performing actions are valid in
all call sites.

jvz committed:
Add localisation

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -425,6 +425,18 @@ private LegitimateButUnknownUserDetails(String username) throws IllegalArgumentE
      */
     @Restricted(NoExternalUse.class)
     public @NonNull Authentication impersonate(@NonNull UserDetails userDetails) {
+        if (!userDetails.isAccountNonLocked()) {
```
It's not a big an oversight as you'd expect because every single implementation of `UserDetails` returns `true` for all four of these methods. I'm working on supporting LDAP and AD specific attributes that would fill in real data in a separate PR for those plugins.

daniel-beck commented:
This looks like something that needs an escape hatch, just in case.

varyvol commented:
@daniel-beck why do you think so? It seems like @jvz will work on supporting LDAP and AD, so if those are saying what should happen with those users, why should it be escaped?

jeffret-b commented:
I need to get a chance to review this and check some related things when I can get the chance.

jeffret-b reviewed (changes requested):
I tested the behavior and it works great. It's a nice simple enhancement to plug this in and get it to work right.

The change really needs an automated test, though. It shouldn't have to rely on behaviors or tests in plugins to verify its proper behavior. It shouldn't be too hard to create an extension of `HudsonPrivateSecurityRealm` that sets these properties differently as needed to test. In `UserTest`, there is already `IdStrategySpecifyingSecurityRealm` that does something similar. Or just extend `SecurityRealm`.

jvz committed:
Merge branch 'master' into JENKINS-55813

jvz committed:
Add tests for JENKINS-55813

Signed-off-by: Matt Sicker <boards@gmail.com>

jeffret-b commented:
Love the new test!

I'd also like to have one for API tokens. If possible it would be really nice to have one for SSH connections. Not sure how difficult that would be. These would really help with regressions and would clarify earlier questions I had.

jvz committed:
[JENKINS-55813] Add API token and SSH CLI tests

This ensures relevant authentication features besides normal web authentication all work as expected (i.e., using User::impersonate).

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz commented:
@jeffret-b added some proper tests to verify those situations.

jeffret-b reviewed (approved):
Looks great!

jvz commented:
Marking this on hold. After experimenting more with LDAP, I've come to the conclusion that these four `UserDetails` attributes are all supposed to be checked by the LDAP server (or equivalent) which will return authentication failure appropriately. These attributes seem to be specific to Spring Security when using a JDBC store for authentication which Jenkins does not support. The Jenkins user database itself does not store metadata for those attributes either, so there's no reason to even check these attributes.

I'm going to try to confirm this hypothesis with more directory servers. One piece of supporting evidence to this whole situation is that I have yet to find an open source project in the wild that checks these attributes of an LDAP account other than LDAP servers themselves. For example, looking for the [policy code in Apache LDAP](https://github.com/apache/directory-server/search?q=in%3Afile+PasswordPolicyConfiguration), it seems as though this is automatically checked as part of a default authentication interceptor configuration. Looking at the [OpenLDAP overlay](https://git.openldap.org/openldap/openldap/-/blob/master/servers/slapd/overlays/ppolicy.c), it seems that OpenLDAP might be doing the same check only in bind operations. While it's possible there might be a way to look up the ppolicy attributes for accounts, it seems as though that should only be done by the LDAP server itself.

jvz commented:
Re-opening this as it's still an improvement over the existing behavior. See https://github.com/jenkinsci/ldap-plugin/pull/50 for associated improvements.

jvz committed:
Add escape hatch

Signed-off-by: Matt Sicker <boards@gmail.com>

jeffret-b commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -142,6 +142,21 @@
+     * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-55813">JENKINS-55813</a>
+     */
+    @Restricted(NoExternalUse.class)
+    public static boolean ALLOW_IMPERSONATION_OF_DISABLED_USERS = SystemProperties.getBoolean(User.class.getName() + ".allowImpersonationOfDisabledUsers");
```
```suggestion
    private static boolean ALLOW_IMPERSONATION_OF_DISABLED_USERS = SystemProperties.getBoolean(User.class.getName() + ".allowImpersonationOfDisabledUsers");
```

jvz committed:
Update core/src/main/java/hudson/model/User.java

Co-authored-by: Jeff Thompson <37345299+jeffret-b@users.noreply.github.com>

oleg-nenashev commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -142,6 +142,21 @@
+     * @see <a href="https://issues.jenkins-ci.org/browse/JENKINS-55813">JENKINS-55813</a>
+     */
+    @Restricted(NoExternalUse.class)
+    private static boolean ALLOW_IMPERSONATION_OF_DISABLED_USERS = SystemProperties.getBoolean(User.class.getName() + ".allowImpersonationOfDisabledUsers");
```
Please create a downstream PR to https://www.jenkins.io/doc/book/managing/system-properties/

jglick replied:
This is not scalable. We need to generate that sort of documentation from sources.

jvz replied:
https://github.com/jenkins-infra/jenkins.io/pull/3953

oleg-nenashev reviewed (approved):
The changelog entry is confusing. Security Plugin implementations can already supply the data. It is a Security Hardening change in Jenkins, because the Jenkins core starts consulting with this data provided by security realms. I suggest re-classifying it as RFE. 

It will also likely require re-testing with https://github.com/jenkinsci/jenkins/pull/4848 ,to ensure alignment with security realms.

jglick commented on test/src/test/java/hudson/model/UserTest.java:
```diff
@@ -24,36 +24,19 @@
-import hudson.security.HudsonPrivateSecurityRealm;
-import hudson.security.Permission;
-import hudson.security.UserMayOrMayNotExistException;
+import hudson.security.*;
```
No wildcards please.

jglick commented on test/src/test/java/hudson/model/UserTest.java:
```diff
@@ -783,4 +773,186 @@ public WebResponse getResponse(WebRequest request) throws IOException {
 
+    @Issue("JENKINS-55813")
+    private static class DisabledAccountsRealm extends AbstractPasswordBasedSecurityRealm {
+        private static class UserStub implements UserDetails {
```
You do not need this stub. Just use `org.acegisecurity.userdetails.User`. That will simplify the merge with #4848.

jglick commented on test/src/test/java/hudson/model/UserTest.java:
```diff
@@ -783,4 +773,186 @@ public WebResponse getResponse(WebRequest request) throws IOException {
+        assertNotNull("Expected API tokens to work for enabled users",
+                j.createWebClient().withBasicApiToken(user).getPage(j.jenkins));
+
+        SSHD sshd = SSHD.get();
+        sshd.setPort(0);
+        sshd.start();
+        try (SshClient client = SshClient.setUpDefaultClient()) {
+            client.setServerKeyVerifier(AcceptAllServerKeyVerifier.INSTANCE);
+            client.start();
+            ConnectFuture future = client.connect("enabled", new InetSocketAddress(sshd.getActualPort()));
+            try (ClientSession session = future.verify(10, TimeUnit.SECONDS).getSession()) {
+                session.addPublicKeyIdentity(generateUserSshKey(user));
+                assertTrue("Expected successful SSH authentication", session.auth().await(10, TimeUnit.SECONDS));
```
It seems you are really testing behaviors of `ssh-cli-auth` and/or `sshd` here? Best to remove this part of the test, or leave it as an integration test in one of those modules, which will also be cleaner when I resurrect #3988 in simplified form.

jvz replied:
This test was added at @jeffret-b's request since the underlying issue related to API tokens and SSH keys. However, if I migrate this functionality all to the LDAP plugin specifically, then I could move these tests there.

jglick replied:
Or to `sshd-module`.

jvz replied:
Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

jvz replied:
I've added a comment in these two methods pointing out that the tests are common authentication mechanisms (you can see web login, API token login, and SSH login).

jglick replied:
> Then that requires `sshd-module` to depend on a version of Jenkins that contains this patch?

Yes, which you can do temporarily using an incremental version, then switch to a weekly if and when this is released.

(Remember that there is no reason to hold back the `jenkins.version` in a `jenkins-module` since they are not distributed to users via the UC.)

jeffret-b replied:
API tokens are implemented in core so the test for it needs to be in core. This is probably as reasonable a place for it as anywhere. Putting the ssh test in sshd-module would be a better location.

jvz replied:
I was not aware about that regarding modules. Now that I'm a maintainer of sshd-module, that's not even too difficult to move this to. Good idea.

jvz replied:
Tests moved to https://github.com/jenkinsci/sshd-module/pull/39 which also discovered a bug in this test as it is.

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -142,6 +142,21 @@
     public static boolean ALLOW_NON_EXISTENT_USER_TO_LOGIN = SystemProperties.getBoolean(User.class.getName() + ".allowNonExistentUserToLogin");
 
+    /**
+     * Jenkins does not allow users to be {@linkplain #impersonate(UserDetails)} impersonated} if its UserDetails
```
```suggestion
     * Jenkins does not allow users to be {@linkplain #impersonate(UserDetails)} impersonated} if its {@link UserDetails}
```

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -142,6 +142,21 @@
+     * return false for any of its validity attributes. These include {@link UserDetails#isEnabled()},
+     * {@link UserDetails#isAccountNonLocked()}, {@link UserDetails#isAccountNonExpired()}, and
+     * {@link UserDetails#isCredentialsNonExpired()}. These attributes are normally enforced either by the underlying
+     * {@link SecurityRealm} service or by the AuthenticationManager. Some realms return user details for these types
```
```suggestion
     * {@link SecurityRealm} service or by the {@link AuthenticationManager}. Some realms return user details for these types
```

jvz committed:
Apply suggestions from code review

Co-authored-by: Jesse Glick <jglick@cloudbees.com>

jvz committed:
Fix javadoc link

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Clean up imports and test code

Signed-off-by: Matt Sicker <boards@gmail.com>

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -425,6 +440,20 @@ private LegitimateButUnknownUserDetails(String username) throws IllegalArgumentE
     public @NonNull Authentication impersonate(@NonNull UserDetails userDetails) {
+        if (!ALLOW_IMPERSONATION_OF_DISABLED_USERS) {
+            if (!userDetails.isAccountNonLocked()) {
+                throw new UsernameNotFoundException(Messages.User_Locked(userDetails.getUsername()));
```
Still trying to figure out in https://github.com/jenkinsci/ldap-plugin/pull/50#issuecomment-707216245 why this should be done in core rather than the plugin defining the `SecurityRealm`.

jvz replied:
Looking more closely at the code changes here, it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself since that will be checked both by `User.getUserDetailsForImpersonation()` and therefore by `User.impersonate()` as well as the few places in Core that call the `impersonate(UserDetails)` variant (those call `getUserDetailsForImpersonation()` first). Being that the `impersonate(UserDetails)` method is restricted from external use, it seems safe to move this downward.

jglick replied:
> it seems like these exceptions could potentially be thrown in the LDAP `UserDetailsService` itself

Right, that was my suggestion. The core PR might still be useful hardening in case there were some unknown realm out there which actually did return an invalid `Authentication`, but we would not be relying on the core check.

jvz replied:
I've already reviewed all the `SecurityRealm` implementations for Jenkins in the wild (and proprietary ones) and found that none of them specify values other than `true` for these four properties. As I explained in greater detail in the LDAP plugin PR, it seems as though these properties are in the Spring Security API mostly for supporting [JDBC authentication providers](https://docs.spring.io/spring-security/site/docs/5.4.1/reference/html5/#servlet-authentication-jdbc) or other embedded user databases (e.g., Jenkins' own internal user database could optionally support these properties).

timja commented:
it can't be on-hold and ready for merge.

@jeffret-b / @jvz what's the reason for on-hold? is it moving the test elsewhere or the system properties change? or something else?

jvz commented:
> what's the reason for on-hold? is it moving the test elsewhere or the system properties change? or something else?

Two major things:

0. Wait for Spring Security upgrade to be merged first to avoid additional merge conflicts.
1. Move the attribute check logic into the underlying `UserDetailsService.loadUserByUsername()` used in specific security realms (like in the LDAP PR) in such a way to avoid any need to patch core.

I'll end up migrating some code from this PR into the others, but I think this PR is stalled until then.

jvz committed:
Remove unnecessary tests

These were moved to sshd-module directly.

Signed-off-by: Matt Sicker <boards@gmail.com>

jvz committed:
Merge branch 'master' into JENKINS-55813

jvz commented:
Based on further testing, it makes sense to include this check in core along with in the affected plugins. I've added a docs update for the system property as well.

oleg-nenashev commented:
I do not see what blocks the merge here.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

jglick commented:
> it makes sense to include this check in core along with in the affected plugins

I see https://github.com/jenkinsci/ldap-plugin/pull/59 was merged, but I am still wondering why we would not just treat this as an implementation detail of `LDAPSecurityRealm`: if there is something wrong with the LDAP entry, throw an exception and refuse to authenticate, and ignore these parts of the Spring Security API as extraneous.

Anyway, no strong opinion, seems harmless to do it this way, just somewhat more complicated.

jvz commented:
It's unexpected behavior that when a user says it's disabled, it is still allowed to use SSH keys or API tokens basically.

jglick commented:
Agreed, I just thought that could be handled (along with GUI login) in `ldap-plugin` without this or https://github.com/jenkinsci/sshd-module/pull/39, by throwing authentication exceptions from suitable methods if the LDAP server indicates the user is disabled, rather than creating `Authentication` & `User` objects but setting various flags on them. Perhaps I misunderstood some subtlety.

jvz commented:
The subtlety is due to API tokens being in core. Also, every other usage of `User.impersonate()` appears to expect a valid user, so this clears up previously ambiguous behavior. SSH keys can be handled in the sshd-module, but API tokens are core.

jvz commented:
> I see [jenkinsci/ldap-plugin#59](https://github.com/jenkinsci/ldap-plugin/pull/59) was merged, but I am still wondering why we would not just treat this as an implementation detail of `LDAPSecurityRealm`: if there is something wrong with the LDAP entry, throw an exception and refuse to authenticate, and ignore these parts of the Spring Security API as extraneous.

Because `User.impersonate()` does _not_ call authenticate. It only looks up user details, and if I threw an error at this point, it will end up using `LegitimateButUnknownUserDetails` which would allow for bypassing the validity check!

jglick commented:
> It only looks up user details, and if I threw an error at this point, it will end up using `LegitimateButUnknownUserDetails`

Only if the `UserDetailsService.loadUserByUsername` implementation throws `UserMayOrMayNotExistException2`. If you throw some other `AuthenticationException`, such as `AccountExpiredException`, from `LDAPUserDetailsService` instead of calling `setAccountNonExpired` and the like, then you will get the desired secure behavior AFAICT: `BasicHeaderApiTokenAuthenticator` will throw this up and prevent the API token from being used.

jglick commented:
Created https://github.com/jenkinsci/ldap-plugin/pull/60 as a demonstration of what I mean, since it seems my earlier comments were not concrete enough.

Removing `ready-for-merge` since discussion is still active.

jvz commented:
Ah, I see spring-security has a nifty new feature. That might work better. I'll look into that and see if I can close this PR.

jglick commented:
It is not a new feature really (would have applied equally to Acegi Security as far as I know), just a different approach to the same problem.

jvz commented:
Well, from what I recall, there were several caller sites that only catch `UsernameNotFoundException` and not the other four Acegi or Spring Security exception variants. Assuming an uncaught exception in those scenarios only leads to an ugly 500 page, that can be cleaned up as discovered.

jvz commented:
Holding this as an alternative for now, though I think the alternate approach in https://github.com/jenkinsci/ldap-plugin/pull/61 should work well enough.

jvz commented:
This PR is no longer needed due to changes in how we're handling this in the affected plugins directly.
